crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1997 @@
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn sha256_num<T: ToBytes>(input: T) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input.to_be_bytes());
+
+    hasher.finalize()
+}
+
+fn sha256_bytes(input: &[u8]) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input);
+
+    hasher.finalize()
+}
+
+fn internal_hash(left_hash: &Hash, right_hash: &Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    hasher.finalize()
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "Side", eq, eq_int))]
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub enum Side {
+    Left,
+    Right,
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "InsertLocation"))]
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub enum InsertLocation {
+    // error: Unit variant `Auto` is not yet supported in a complex enum
+    // = help: change to a struct variant with no fields: `Auto { }`
+    // = note: the enum is complex because of non-unit variant `Leaf`
+    Auto {},
+    AsRoot {},
+    Leaf { index: TreeIndex, side: Side },
+}
+
+const NULL_PARENT: TreeIndex = 0xffff_ffffu32;
+
+#[derive(Debug, PartialEq)]
+pub struct NodeMetadata {
+    pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+impl NodeMetadata {
+    pub fn from_bytes(blob: MetadataBytes) -> Result<Self, String> {
+        // OPT: could save 1-2% of tree space by packing (and maybe don't do that)
+        Ok(Self {
+            node_type: Self::node_type_from_bytes(blob)?,
+            dirty: Self::dirty_from_bytes(blob)?,
+        })
+    }
+
+    pub fn to_bytes(&self) -> MetadataBytes {
+        let mut bytes = [0u8; METADATA_SIZE];
+        bytes[TYPE_RANGE].copy_from_slice(&[self.node_type.to_u8()]);
+        bytes[DIRTY_RANGE].copy_from_slice(&[u8::from(self.dirty)]);
+
+        bytes
+    }
+
+    pub fn node_type_from_bytes(blob: MetadataBytes) -> Result<NodeType, String> {
+        NodeType::from_u8(u8::from_be_bytes(blob[TYPE_RANGE].try_into().unwrap()))
+    }
+
+    pub fn dirty_from_bytes(blob: MetadataBytes) -> Result<bool, String> {
+        match u8::from_be_bytes(blob[DIRTY_RANGE].try_into().unwrap()) {
+            0 => Ok(false),
+            1 => Ok(true),
+            other => Err(format!("invalid dirty value: {other}")),
+        }
+    }
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "Node", get_all))]
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    parent: Parent,
+    hash: Hash,
+    specific: NodeSpecific,
+}
+
+// #[cfg_attr(feature = "py-bindings", pyclass(name = "NodeSpecific"))]
+#[cfg_attr(feature = "py-bindings", pyclass(name = "NodeSpecific", get_all))]
+#[derive(Clone, Debug, PartialEq)]
+pub enum NodeSpecific {
+    Internal { left: TreeIndex, right: TreeIndex },
+    Leaf { key: KvId, value: KvId },
+}
+
+impl NodeSpecific {
+    // TODO: methods that only handle one variant seem kinda smelly to me, am i right?
```

Comment:
Yes these enum variants should be factored out into structs if you need to add special methods on one variant.